python/ray/dag/compiled_dag_node.py

```diff
@@ -895,28 +917,56 @@ def _get_or_compile(
-                    )
+                    for output_task, readers in output_to_readers.items():
+                        reader_and_node_list: List[
+                            Tuple["ray.actor.ActorHandle", str]
+                        ] = []
+                        for reader in readers:
+                            reader_and_node_list.append(
+                                (
+                                    self._dag_creator_actor,
+                                    self._get_node_id(self._dag_creator_actor),
+                                )
+                            )
+                        output_task_to_reader_and_node_list[
+                            output_task
+                        ] = reader_and_node_list
                 else:
-                    for reader in readers:
-                        reader_handle = reader.dag_node._get_actor_handle()
-                        reader_and_node_list.append(
-                            (reader_handle, self._get_node_id(reader_handle))
-                        )
+                    for output_task, readers in output_to_readers.items():
+                        reader_and_node_list: List[
+                            Tuple["ray.actor.ActorHandle", str]
+                        ] = []
+                        for reader in readers:
+                            reader_handle = reader.dag_node._get_actor_handle()
+                            reader_and_node_list.append(
+                                (reader_handle, self._get_node_id(reader_handle))
+                            )
+                        output_task_to_reader_and_node_list[
+                            output_task
+                        ] = reader_and_node_list
 
                 fn = task.dag_node._get_remote_method("__ray_call__")
-                task.output_channel = ray.get(
-                    fn.remote(
-                        do_allocate_channel,
-                        reader_and_node_list,
-                        typ=type_hint,
+                for (
+                    output_task,
+                    reader_and_node_list,
+                ) in output_task_to_reader_and_node_list.items():
+                    output_channel = ray.get(
+                        fn.remote(
+                            do_allocate_channel,
+                            reader_and_node_list,
+                            typ=type_hint,
+                        )
                     )
-                )
+                    output_idx = None
+                    output_task_node = output_task.dag_node
+                    if isinstance(output_task_node, ClassMethodOutputNode):
+                        output_idx = output_task_node.output_idx
+                        output_task.output_channels.append(output_channel)
+                        output_task.output_idxs.append(output_idx)
+                    task.output_channels.append(output_channel)
+                    task.output_idxs.append(output_idx)
```

Comment:
I got a bit confused by this code; what is the difference between `output_task` and `task`, and why do we only sometimes add to `output_task.output_channels`?

Reply:
`output_task` are downstream `ClassMethodOutputNode` of a `ClassMethodNode`. Since we only allocate output channels at the `ClassMethodNode`, the downstream `ClassMethodOutputNode`s have to copy the allocated output channels. `output_task` only copy output channels when it is the `ClassMethodOutputNode`.

Reply:
`output_task` is the same `output_task` in `output_task_to_readers` and `output_task_to_reader_and_node_list`. I add some comments to make them clear.

Reply:
Now `ClassMethodOutputNode` is a `ClassMethodNode` with `self.is_class_method_output == True`.